README.md

```diff
@@ -1 +1,122 @@
+* `consume_ctx` controls how the parsed results are consumed (e.g. inserted into a database, appended to a `DataFrame`...). `consume_ctx` allows the user to dispatch on their `consume!` method and hold any state necessary for consumption. This happens immediately after `populate_result_buffer!`.
+* `chunking_ctx` controls how the work on individual chunks of data is scheduled. It contains buffers for input bytes and found newlines. Through this struct the user controls the size of the chunks and the number of spawned tasks that carry out the parsing and consuming. If there is enough data in the input, a secondary `chunking_ctx` is created internally to facilitate the double-buffering described above.
+* `result_buffers` controls in which format the results are stored. These result buffers hold results from `populate_result_buffer!` and are passed to `consume!`. This allows the user to have multiple result formats for the with `parsing_ctx` e.g. row oriented vs column oriented buffers.
+
+See the docstring of `populate_result_buffer!` and `consume!` for more information about how to integrate with them.
+
+## Example: Examining the results of the Lexer
+
+Since `ChunkedBase.jl` handles newline detection automatically, a very simple chunked processor could just show the newline positions to the user. Of course, this is not very useful apart from demonstrating how to use this package.
+
+```julia
+using ChunkedBase
+using NewlineLexers: Lexer
+
+# Our chunked processor doesn't require any settings nor does it need to maintain
+# any additional state, so we'll define our `ParsingContext` and `ConsumeContext` only
+# for dispatch reasons.
+struct ParsingContext <: AbstractParsingContext end
+struct ConsumeContext <: AbstractConsumeContext end
+# Our buffer will hold the newline positions in the current chunk
+struct ResultBuffer <: AbstractResultBuffer
+    newlines_in_chunk::Vector{Int}
+end
+
+# Our overload copies `newlines_segment` to our result buffer
+function ChunkedBase.populate_result_buffer!(
+    result_buf::ResultBuffer,
+    newlines_segment::AbstractVector{Int32},
+    ::ParsingContext,
+    ::Vector{UInt8},
+    ::Union{Nothing,Vector{UInt8}}=nothing,
+    ::Type=Tuple{}
+)
+    resize!(result_buf.newlines_in_chunk, length(newlines_segment))
+    result_buf.newlines_in_chunk .= newlines_segment
+    return nothing
+end
+
+# We consume result buffer by simply printing it
+function ChunkedBase.consume!(::ConsumeContext, payload::ParsedPayload)
+    chunking_ctx = payload.chunking_ctx
+    result_buffer = payload.results
+
+    # To demonstrate how double-buffering works, print the buffer id and the refill number
+    chunk_info = (chunking_ctx.id, chunking_ctx.buffer_refills[])
```

Comment:
there's a lot of accessing the fields of types... are these all public API? what type even is this? oh `ChunkingContext` is a concrete type that users have to use, whereas everything else they define their own types? i think i need more examples / detail on all these things

Reply:
Yes these are public. I was hoping that in this example, the object and field names together with comments would be understandable, since you can see everything being constructed, but it seems you are missing some information... Do you know what makes it confusing for you?